internal/dot/graph.go

```diff
@@ -60,19 +81,29 @@ type Group struct {
 	Type    reflect.Type
 	Name    string
 	Results []*Result
+	State   FailureState
 }
 
 // Graph is the DOT-format graph in a Container.
 type Graph struct {
-	Ctors  []*Ctor
-	Groups map[groupKey]*Group
+	Ctors   []*Ctor
+	ctorMap map[CtorID]*Ctor
+
+	Groups   []*Group
+	groupMap map[groupKey]*Group
+
+	Failed *failedNodes
 }
 
-// Node is a single node in a graph and is embedded into Params and Results.
-type Node struct {
-	Type  reflect.Type
-	Name  string
-	Group string
+type failedNodes struct {
+	// RootCauses is a list of the point of failures. They are the root causes
+	// of failed invokes and can be either missing types (not provided) or
+	// error types (error providing).
+	RootCauses []*Result
```

Comment:
Will this ever have a length more than one? It'll either be zero or one, right?

This is fine for now but in a future change we probably can simplify the internal representation to be,

    // Results that failed to build, in-order. The first result is the root
    // cause, the second result failed to build because the first was
    // unavailable, and so on.
    FailureChain []*Result

Reply:
+1 here - it was fairly confusing at first glance to see a slice of roots.

Reply:
There will be multiple root causes in the case of multiple missing types (`errMissingManyTypes`). IMO it would only make sense if all of them are marked red.